dsp/transform/hilbert.go

```diff
@@ -0,0 +1,83 @@
+// Copyright ©2024 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package transform
+
+import (
+	"gonum.org/v1/gonum/dsp/fourier"
+)
+
+// Hilbert implements an approximate Hilbert transform that allows calculation
+// of an approximate analytical signal of a real signal, and determine the
+// real envelope of a signal.
+type Hilbert struct {
+	fft  *fourier.CmplxFFT
+	work []complex128
+}
+
+// NewHilbert returns a new Hilbert transformer for signals of size n.
+// The transform is most efficient when n is a product of small primes.
+func NewHilbert(n int) *Hilbert {
+	if n < 1 {
+		return &Hilbert{nil, nil}
+	}
+	return &Hilbert{
+		fft:  fourier.NewCmplxFFT(n),
+		work: make([]complex128, n),
+	}
+}
+
+// Len returns the length of signals that are valid input for this Hilbert transform.
+func (h *Hilbert) Len() int {
+	return len(h.work)
+}
+
+// AnalyticSignal computes the analytical signal of a real signal, and stores
+// the result in the dst slice, returning it.
+//
+// If the dst slice is nil, a new slice will be created and returned. The dst slice
+// must be the same length as the input signal, otherwise the method will panic.
+func (h *Hilbert) AnalyticSignal(dst []complex128, signal []float64) []complex128 {
+	if len(signal) != h.Len() {
+		panic("transform: input signal length mismatch")
+	}
```

Comment:
I was trying to keep this struct's API similar to the FFT API, which does not return `error`s for its compute methods, preferring to panic. The constraint is noted in the doc comment, and I can't think of a scenario where someone would reasonably expect to get an analytic signal with different size from the input, so panic felt more appropriate anyway (implementation rather than user error).

But if `error` is the desired API here, I can put it in.

Reply:
> I would change signature to return an error and avoid panics

No. Please see https://go-talks.appspot.com/github.com/gonum/talks/2017/gonumtour.slide#14

Reply:
OK.